### PR TITLE
Update dev docs

### LINKED ISSUE #11354: Update developer docs
The developer docs have grown organically with Bokeh over many years. Some parts of them are incomplete or now outdated. To make the developer tools a more useful resource for existing and new contributors, we will update, edit, and restructure the existing information. This is part of the maintenance and sustainability work for the CZI R3 grant (#11351).

### PR DESCRIPTION
**Pull request by tcmetzger**
This PR supersedes #11418 and collects my updates to the developer docs. Instead of trying to have all updates in one PR, I'm trying to split up the work to make it more manageable. This PR contains the following updated/new chapters:

- [x] Landing page
- [x] Contributing to the documentation
- [x] Running tests
- [x] Writing tests
- [x] Creating issues
- [x] Creating a successful pull request

The following three chapters will be updated in subsequent pull requests: "Working with Bokeh models", "Understanding Bokeh server", "Contributing to BokehJS". These three chapters are currently in their original state, with only small changes to adapt to the new structure. This way, the developer docs remain complete and functional, even before merging any subsequent PRs.

Fixes #11354

### CODE REVIEW

File: sphinx/source/docs/dev_guide/writing_tests.rst [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+    instead of one complex one.
+
+Use the |bokeh.models| API whenever possible
+    Try to use Bokeh's
+    :ref:`low-level bokeh.models interface <userguide_interfaces_models>` instead of
+    the more high-level
+    :ref:`bokeh.plotting interface <userguide_interfaces_plotting>`.
+
+After adding or updating an integration test, you need to create a new baseline
+image. To create a new base image, add ``--set-new-base-screenshot`` to the
```
Comment (tcmetzger) [inline sphinx/source/docs/dev_guide/writing_tests.rst:81]: @mattpap Is it possible that the current docs are outdated here?  I just get  `pytest: error: unrecognized arguments: --set-new-base-screenshot` and I couldn't find anywhere in the code that would handle this argument.

File: sphinx/source/docs/dev_guide/writing_tests.rst [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+attributes. ``equal_attributes`` accepts an array of strings as an optional
+``ignored_attributes`` parameter.
+
+For example:
+
+.. code-block:: TypeScript
+
+    expect_element(svg).to.have.equal_attributes(string_to_html(`<p>test</p>`))
+
+If you add new test files, you should add an entry in the directory index file.
```
Comment (tcmetzger) [inline sphinx/source/docs/dev_guide/writing_tests.rst:169]: @mattpap Could you provide some background here? I was not able to find a file named `ìndex` in `bokehjs/tests/unit` folder. Or is it possible that the index file refers to non-unit tests for BokehJS, such as defaults and codebases tests that do have an `index.ts` file?
For reference, this is the original sentence in the current docs: https://docs.bokeh.org/en/latest/docs/dev_guide/testing.html#javascript-unit-tests

File: sphinx/source/docs/dev_guide/setup.rst [hunk 3]
```diff
@@ -249,45 +278,144 @@ With either mode, you will be prompted for how to install BokehJS, e.g.:
...
+``BOKEH_RESOURCES``
+~~~~~~~~~~~~~~~~~~~
+
+When working on Bokeh's codebase, the most important environment variable to be
+aware of is ``BOKEH_RESOURCES``. This variable controls which version of
+:term:`BokehJS` to use.
+
+By default, Bokeh downloads any necessary JavaScript code for BokehJS from a
+Content Delivery Network (CDN). If you want Bokeh to use your local BokehJS
+version instead, you should set ``BOKEH_RESOURCES`` to ``absolute-dev``:
```
Comment (tcmetzger) [inline sphinx/source/docs/dev_guide/setup.rst:342]: @hyles-lineata @bryevdv @mattpap This refers to the issue raised in #10694. Basically, the question is: What should we recommend for someone who generally wants to use a local BokehJS? When setting `BOKEH_DEV` to `true`, the variable ``BOKEH_RESOURCES`` will be set to ``absolute-dev``. This is why I assumed `absolute-dev` to be the best default value in this context.

However, I have mostly used `inline` in the past, same as Bryan. Matteusz recommends `server-dev`.

The paragraph above links to `bokeh.settings` which contains all information about these variables. I will also explain `BOKEH_RESOURCES` a little more in depth the upcoming BokehJS section of the dev guide, as Mateusz suggested.

File: sphinx/source/docs/dev_guide/testing.rst [hunk 4]
```diff
@@ -1,231 +1,447 @@
...
-~~~~~~~~~~~~~
+        pytest tests/integration
 
-To run any of the tests with coverage, use the following:
+    These tests mostly simulate UI interactions. Therefore, they require
+    `Chrome`_ or `Chromium`_ and `Selenium`_ with the `ChromeDriver`_ web
+    driver.
 
-.. code-block:: sh
+Code coverage
```
Comment (tcmetzger) [inline sphinx/source/docs/dev_guide/testing.rst:131]: @bryevdv @mattpap I was not able to find many details about coverage in the current docs. Do we have a certain coverage (x%) that we are aiming for? Or is this mostly just for informational purposes?

File: sphinx/source/docs/dev_guide/setup.rst (hunk 3 above)
Comment (bryevdv) [reply sphinx/source/docs/dev_guide/setup.rst:342]: I don't think we are going to agree here on the advice here. So I'd actually just punt and be direct:

> By default, Bokeh loads latest published BokehJS from CDN. If you have modified and BokehJS and built it locally, you will need to make a configuration change in order to actually use the local version with your changes. Here are three ways to use local BokehJS:

And then explain all three.

File: sphinx/source/docs/dev_guide/testing.rst (hunk 4 above)
Comment (bryevdv) [reply sphinx/source/docs/dev_guide/testing.rst:131]: Coverage is tricky since we have lots of different kinds of tests and the coverage tool(s) only work with some of them The instructions here are only for the Python unit tests, which hover around 90% and I would like to keep it that way. There is some way to also get JS unit test coverage but I don't remember what it is. We've never looked at it with any regularity. The would be two completely separate reports which is also a pain. In principle I think the different coverage tools can all output XML reports that  could potentially be merged in a post-processing step, but no one has ever looked into it. For the selenium, examples, visual js tests, etc. I'm not sure if it is possible to get coverage info at all.

TLDR; Wnat ot keep Python unit test coverage around 90+% we should keep information about that if nothing else.

**Comment by tcmetzger** [comment]
Currently, all filenames (rst) are unchanged to make sure we don't get any dead links. Files that are no longer needed (such as `bindings.rst` that is now included in the landing page) still exist but contain a link to the new location of the content, together with `:orphan:`.

File: sphinx/source/docs/dev_guide/testing.rst (hunk 4 above)
Comment (tcmetzger) [reply sphinx/source/docs/dev_guide/testing.rst]: Sounds good! I updated the section to clarify that this only applies to Python unit tests and that the goal is around 90% coverage there.

File: sphinx/source/docs/dev_guide.rst [hunk 5]
```diff
@@ -1,64 +1,264 @@
...
+  was started by `@hafen <hafen_>`_.
+* `bokeh.jl <bokeh_jl_>`_ brings Bokeh to Julia users. It was originally created
+  by `@samuelcolvin <samuelcolvin_>`_.
+* `bokeh-scala <bokeh_scala_>`_ exposes Bokeh in the Scala language. The
+  project was created by the Bokeh core dev team member `@mattpap <mattpap_>`_.
+
+The low-level object interface in Python mirrors the JSON schema exactly.
+Therefore, the best, most authoritative source of information for anyone
+writing bindings for Bokeh are the reference guide sections for the
+:ref:`bokeh.core.properties` and |bokeh.models|.
```
Comment (bryevdv) [inline sphinx/source/docs/dev_guide.rst:218]: I suppose we could mention the `spec.py` script here:

https://github.com/bokeh/bokeh/blob/branch-2.4/scripts/spec.py

It will spit out a JSON description of every model and property, including type and default and helpstring, that could be used as part of automated low-level bindings creation.

File: sphinx/source/docs/dev_guide.rst [hunk 6]
```diff
@@ -1,64 +1,264 @@
...
+
+If you would like to start working on a new binding or have a new binding added
+to this list, please contact the Bokeh core team through `Slack`_.
+
+.. _contributors_guide_spread_the_word:
+
+Spread the word
+^^^^^^^^^^^^^^^
+
+Finally, as an open source project, Bokeh relies on word-of-mouth advertising.
```
Comment (bryevdv) [inline sphinx/source/docs/dev_guide.rst:228]:
```suggestion
Finally, as an open source project, Bokeh relies on word-of-mouth to reach new users.
```

File: sphinx/source/docs/dev_guide/creating_pull_requests.rst [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+2. Create a branch
+    Before writing any code, you need to create a new `branch`_ on
+    :ref:`your fork <devguide_setup_cloning>` of the Bokeh repository. In most
+    cases, you should base your new branch on Bokeh's ``default`` branch. The
+    default branch is usually a branch that represents the next version of
+    Bokeh, for example ``branch-2.4``.
+
+    Use this pattern to name your branch:
+
+    ``[issue number]_[Python/BokehJS model you are working on, if applicable]_[optional: model property or short description]``
```
Comment (bryevdv) [inline sphinx/source/docs/dev_guide/creating_pull_requests.rst:46]: I'd personally be OK with
```suggestion
    ``[issue number]_[short description]``
```

File: sphinx/source/docs/dev_guide/creating_pull_requests.rst [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+    For example:
+
+    ``11423_table_column_visible``
+
+    See `BEP 6: Branching Strategy`_ for more information on the different
+    branches in the Bokeh repository.
+
+3. Publish your branch and push a first commit
+    Before you can open a pull request, you first need to publish your branch.
+    At this time, you should also include a first commit.
```
Comment (bryevdv) [inline sphinx/source/docs/dev_guide/creating_pull_requests.rst:57]: I haven't really heard "publish" in this context. I think usually just more like "push the branch from your local clone, to your fork on GitHub" and it would be good to link to GH docs or tutorials about that I think (or an example command)

File: sphinx/source/docs/dev_guide/creating_pull_requests.rst [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+    to the issue your pull request is based on. To do so, use a keyword such as
+    "fixes," followed by the number of the issue. For example "Fixes #11479".
+    See `Linking a pull request to an issue using a keyword`_ in the
+    `GitHub documentation`_ for more information. Your description should also
+    include information about :ref:`tests <devguide_writing_tests>` and
+    :ref:`documentation <devguide_documentation>`, if applicable.
+
+6. Add more commits to your pull request
+    Once you have created a pull request, keep adding commits. A member of the
+    `Bokeh core team`_ will review your pull request and help you along the way.
```
Comment (bryevdv) [inline sphinx/source/docs/dev_guide/creating_pull_requests.rst:89]: This part is sometimes not necessary. Maybe 
> will review your pull request,  and may request changes or additions. If so, they will help you along the way with any questions you may have. You can make new changes in your existing local branch, and push them to Github. The PR will update automatically (there is no need to open a new PR)

Also I think that last bit is important, some folks definitely try to close a PR just to add new commits.

File: sphinx/source/docs/dev_guide/creating_pull_requests.rst [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+    request and merge your changes into the repository.
+
+7. Look for the next issue to work on
+    With your first pull request merged, you should take another look at Bokeh's
+    `issue tracker`_ to find the next issue to work on.
+
+    Once your successfully complete two substantive pull requests, you are
+    eligible to become a member of the `Bokeh development team`_. See
+    `BEP 4: Project Roles`_ for more information about all roles in the Bokeh
+    project.
```
Comment (bryevdv) [inline sphinx/source/docs/dev_guide/creating_pull_requests.rst:105]: I'm not sure it's something to mention here but just for record one reason it is good for folks to join `@bokeh/dev` is that then they can work from a clone of the main repo instead of a fork. AFAIK this is still necessary for the examples test report to be able to be published. It also makes it much easier for other core people to check out the branch (and when necessary and appropriate collaborate directly by pushing to the branch)

File: sphinx/source/docs/dev_guide/creating_issues.rst [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+the GitHub repository at `<https://github.com/bokeh/bokeh/issues>`_.
+
+.. _devguide_issues_before:
+
+Before opening an issue
+-----------------------
+
+As a first step, you should always check
+`Bokeh's GitHub issue tracker <issue tracker_>`_ and the
+`Bokeh Discourse`_ for similar or identical issues.
```
Comment (bryevdv) [inline sphinx/source/docs/dev_guide/creating_issues.rst:17]: Maybe an encouraging note:

> There are lots of issues, and it's possible to mis duplicates even after searching. If team member closes your issue as s duplicate, don't worry! Your valuable input can be added to the linked issue. 

?

File: sphinx/source/docs/dev_guide/creating_issues.rst [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+    other packages relevant to the bug you are reporting.
+
+Description of expected behavior and the observed behavior
+    Describe the behavior you would have expected and the behavior you observed
+    instead.
+
+Complete, minimal, self-contained example code that reproduces the issue
+    Please provide a code example that follows the
+    `definition of a minimal, reproducible example at stackoverflow.com`_.
+    For more details, see `Craft Minimal Bug Reports at matthewrocklin.com`_.
```
Comment (bryevdv) [inline sphinx/source/docs/dev_guide/creating_issues.rst:83]: I don't know if there is a good place to fit it in but FWIW it is 1000% better to get MREs as simple scripts and not code to run in notebook, unless the issue is very specifically to do with the notebook in particular. Also screenshots of code are always *very discouraged*

File: sphinx/source/docs/dev_guide/setup.rst [hunk 13]
```diff
@@ -249,45 +278,217 @@ With either mode, you will be prompted for how to install BokehJS, e.g.:
...
+
+This way, Bokeh will use local and unminified BokehJS resources, the default log
+levels are increased, the generated HTML and JSON code will be more
+human-readable, and Bokeh will not open a new browser window each time |show| is
+called.
+
+.. note::
+    Setting ``BOKEH_DEV=true`` and therefore enabling
+    ``BOKEH_RESOURCES=absolute-dev`` causes rendering problems when used
+    with :term:`Bokeh server <Server>` or in
```
Comment (bryevdv) [inline sphinx/source/docs/dev_guide/setup.rst:472]:
```suggestion
    Setting ``BOKEH_DEV=true`` enables ``BOKEH_RESOURCES=absolute-dev``, which 
    can causes rendering problems when used with :term:`Bokeh server <Server>` or in
```

File: sphinx/source/docs/dev_guide/setup.rst [hunk 14]
```diff
@@ -297,65 +498,112 @@ You should see output similar to:
...
-    IPython version     :  7.15.0
-    Tornado version     :  6.0.4
-    Bokeh version       :  2.0.2-95-g8e0b447c0-dirty
-    BokehJS static path :  /Users/bryan/work/bokeh/bokeh/server/static
-    node.js version     :  v14.4.0
-    npm version         :  6.14.5
+    Python version      :  3.9.6 | packaged by conda-forge | (default, Jul 11 2021, 03:39:48)
+    IPython version     :  7.25.0
+    Tornado version     :  6.1
+    Bokeh version       :  2.4.0dev1-42-g9c3ee2f7e-dirty
```
Comment (bryevdv) [inline sphinx/source/docs/dev_guide/setup.rst:504]: we just updated to pep440 version I think this should be
```suggestion
    Bokeh version       :  2.4.0dev1+42.g9c3ee2f7e.dirty
```

File: sphinx/source/docs/dev_guide/setup.rst [hunk 15]
```diff
@@ -297,65 +498,112 @@ You should see output similar to:
...
-If you have any problems with the steps here, please `contact the developers`_.
+redrawing the line with the new parameters. The ``--show`` option opens a
+web browser, the default URL for the Bokeh server is ``localhost:5006``.
 
+.. note::
+    Updating an existing development environment does not always work as
+    expected. If you get errors after updating an older environment, use
+    ``conda remove --name bkdev --all``, delete your local ``bokeh`` folder,
+    and start afresh, following the steps in this guide from
+    :ref:`the beginning <devguide_setup_preliminaries>`.
```
Comment (bryevdv) [inline sphinx/source/docs/dev_guide/setup.rst:581]: Should this note be up in the conda section? It seems odd for it to be in "Run Bokeh Server" Or maybe a dedicated troubleshooting section?

File: sphinx/source/docs/dev_guide/documentation.rst [hunk 16]
```diff
@@ -175,71 +198,145 @@ When building Bokeh's documentation, the most common options for ``make`` are:
...
 called `Napoleon`_ with `Napoleon's Google style`_. For Napoleon to work
-correctly, all docstrings you write should follow the rules in the `Google
+correctly, all docstrings you write need to follow the rules in the `Google
 Python Style Guide`_.
 
 Docstrings for functions and methods generally include these three elements:
 
-* A short description of what the function or method does, starting with a
+* A short description of what the function does, starting with a
   verb. For example: "Creates and returns a new Foo."
```
Comment (bryevdv) [inline sphinx/source/docs/dev_guide/documentation.rst:337]:
```suggestion
  verb. For example: "Create and return a new Foo."
```
would be consistent with the voice I've tried to use throughout since the beginning

**Comment by tcmetzger** [comment]
@mattpap as discussed, I'd appreciate your feedback, especially on the two chapters Running tests (testing.rst) and Writing tests (writing_tests.rst). Thank you!

File: sphinx/source/docs/dev_guide/writing_tests.rst [hunk 17]
```diff
@@ -0,0 +1,367 @@
...
+Use ``expect()`` together with the following elements to create assertions for
+the BokehJS testing framework:
+
+* ``to`` and ``be``: tokens to improve readability of assertions and connect
+  elements
+* ``not``: negates the following assertions
+* ``throw``: asserts that an error is thrown. Accepts the following optional
+  parameters: ``error_type`` (filter by ``Error``) and ``pattern`` (filter by
+  regular expression or string).
+* ``equal``: asserts strict equality (``===``). Expects an operand to compare
```
Comment (mattpap) [inline sphinx/source/docs/dev_guide/writing_tests.rst:135]: Equality is defined as (deep) value equality, instead of reference equality given by `==` or `===`. For reference equality one would use `identical` method (strict, i.e. `===`). Value equality is provided by `core/util/eq` module, thus is the same across bokehjs (e.g. properties use it to detect change).

File: sphinx/source/docs/dev_guide/writing_tests.rst [hunk 18]
```diff
@@ -0,0 +1,367 @@
...
+  to.
+* ``similar``: asserts similarity within a defined tolerance. Expects an operand
```
Comment (mattpap) [inline sphinx/source/docs/dev_guide/writing_tests.rst:137]: This uses the same value equality as `equal` method, with the exception of numbers.

File: sphinx/source/docs/dev_guide/writing_tests.rst [hunk 19]
```diff
@@ -0,0 +1,367 @@
...
+  to compare to as well as an optional ``number`` as ``tolerance``.
+* ``identical``: asserts same-value equality. Expects an operand to compare
```
Comment (mattpap) [inline sphinx/source/docs/dev_guide/writing_tests.rst:139]: As mentioned earlier, this is exactly `===`.

File: sphinx/source/docs/dev_guide/writing_tests.rst [hunk 20]
```diff
@@ -0,0 +1,367 @@
...
+    .. code-block:: TypeScript
+
+        describe("Your Object", () => {
+        it("should show certain behavior", async () => {
+            const p = fig([width, height], {figure_attrs})
+
+            ...
+
+            await display(p)
+        })
```
Comment (mattpap) [inline sphinx/source/docs/dev_guide/writing_tests.rst:250]: Indentation is off here and a few other examples.

**Comment by mattpap** [comment]
@tcmetzger, I would mention the special nature of regression tests (both unit and integration) (how descriptions should be formulated, including issue numbers). Also it would be worth mentioning that all bug fixes and especially regression fixes must come with a regression tests of either or both kinds.

**Comment by mattpap** [comment]
I would also mention about `assert()` vs `expect()`, especially when dealing with types, e.g.:
https://github.com/bokeh/bokeh/blob/79903c70959cddcf26abb5b91e32b80b8e143fbf/bokehjs/test/unit/regressions.ts#L116-L118
i.e. `assert()` allows to narrow types without need for presumptuous casting.

**Comment by tcmetzger** [comment]
I've added all suggestions and labeled this PR as ready. After the 2.4 release, I'll work on the remaining chapters (Models, Server, and BokehJS), and can also add more improvements to the chapters in this PR.

**Comment by mattpap** [comment]
Test failures are unrelated and will by resolved by PR #11566.

File: sphinx/source/docs/dev_guide/writing_tests.rst [hunk 21]
```diff
@@ -0,0 +1,417 @@
...
+``expect_element()`` function. You can use this function in combination with
+``have`` and ``equal_attributes`` to test whether an element has certain
+attributes. ``equal_attributes`` accepts an array of strings as an optional
+``ignored_attributes`` parameter.
+
+For example:
+
+.. code-block:: TypeScript
+
+    expect_element(svg).to.have.equal_attributes(string_to_html(`<p>test</p>`))
```
Comment (mattpap) [inline sphinx/source/docs/dev_guide/writing_tests.rst:190]: This is outedated.

File: sphinx/source/docs/dev_guide/writing_tests.rst [hunk 22]
```diff
@@ -0,0 +1,417 @@
...
+BokehJS unit tests
+~~~~~~~~~~~~~~~~~~
+
+The :term:`BokehJS` unit tests help make sure that the individual sections of
+BokehJS function as expected. The unit tests for BokehJS are located in the
+:bokeh-tree:`bokehjs/test/unit/` folder and sub-folders.
+
+Use `Chai "expect" assertion style <Chai_>`_ when writing unit tests for the
+BokehJS testing framework. See the `API documentation of the Chai Assertion
+Library <Chai documentation_>`_ for more details on this style.
```
Comment (mattpap) [inline sphinx/source/docs/dev_guide/writing_tests.rst:141]: Referring to Chai's documentation may give people a wrong impression regarding bokehjs' assertion framework. Both are quite similar only on the superficial level. If we need to mention Chai at all, then only to say that our framework was inspired by.

**Comment by tcmetzger** [comment]
As mentioned above, the CI failures for macOS/windows are not related to my changes.

File: sphinx/source/docs/dev_guide/setup.rst [hunk 23]
```diff
@@ -279,65 +462,124 @@ You should see output similar to:
...
+:ref:`conda environment <contributor_guide_setup_creating_conda_env>`,
+:ref:`Node packages <contributor_guide_setup_installing_node_packages>`, and
+:ref:`local build <contributor_guide_setup_install_locally>` are up to date.
 
+If you keep geting errors after updating an older environment, use
+``conda remove --name bkdev --all``, delete your local ``bokeh`` folder,
+and start afresh, following the steps in this guide from
+:ref:`the beginning <contributor_guide_setup_preliminaries>`.
+
+For more information on running and installign Bokeh, check the
```
Comment (bryevdv) [inline sphinx/source/docs/dev_guide/setup.rst:554]:
```suggestion
For more information on running and installing Bokeh, check the
```

File: sphinx/source/docs/dev_guide/setup.rst [hunk 24]
```diff
@@ -91,133 +117,119 @@ Then, activate the environment:
...
-If you do not wish to install globally (i.e. with ``-g``), then all
-subsequent ``npm`` commands will need to be adjusted to use the local
+If you do not want to install npm globally, leave out the ``-g`` flag. In this
+case, you need to adjust all subsequent ``npm`` commands to use the local
 version installed under ``bokehjs/node_modules``.
 
-Next, still in the ``bokehjs`` subdirectory, execute the following command
-to install all of BokehJS JavaScript dependencies:
+Next, still in the ``bokehjs`` subdirectory, run the following command
+to install all of BokehJS' JavaScript dependencies:
```
Comment (bryevdv) [inline sphinx/source/docs/dev_guide/setup.rst:154]: extreme nit: I try to avoid trailing possessives like this, they read awkwardly to me in text. Maybe it's just me. But FWIW I think any of these would be preferable

> to install all BokehJS JavaScript dependencies:

> to install all the BokehJS JavaScript dependencies:

> to install all the JavaScript dependencies for BokehJS:

**Comment by bryevdv** [comment]
Unfortunately I only had time to skim this morrning. One type and one small comment for consideration, otherwise LGTM

**Review by mattpap — APPROVED** [review]
